subsys/net/lib/sockets/sockets_can.c

```diff
@@ -0,0 +1,355 @@
+	}
+
+	ret = net_context_sendto_new(ctx, buf, len, dest_addr, addrlen, NULL,
+				     timeout, NULL, ctx->user_data);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	return len;
+}
+
+static ssize_t zcan_recvfrom_ctx(struct net_context *ctx, void *buf,
+				 size_t max_len, int flags,
+				 struct sockaddr *src_addr,
+				 socklen_t *addrlen)
+{
+	size_t recv_len = 0;
+	s32_t timeout = K_FOREVER;
+	struct net_pkt *pkt;
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
+	}
+
+	if (flags & ZSOCK_MSG_PEEK) {
+		int ret;
+
+		ret = _k_fifo_wait_non_empty(&ctx->recv_q, timeout);
+		/* EAGAIN when timeout expired, EINTR when cancelled */
+		if (ret && ret != -EAGAIN && ret != -EINTR) {
+			errno = -ret;
+			return -1;
+		}
+
+		pkt = k_fifo_peek_head(&ctx->recv_q);
+	} else {
+		pkt = k_fifo_get(&ctx->recv_q, timeout);
+	}
+
+	if (!pkt) {
+		errno = EAGAIN;
+		return -1;
+	}
+
+	/* We do not handle any headers here, just pass the whole packet to
+	 * the caller.
+	 */
+	recv_len = net_pkt_get_len(pkt);
+	if (recv_len > max_len) {
+		recv_len = max_len;
+	}
+
+	if (net_pkt_read_new(pkt, buf, recv_len)) {
+		errno = EIO;
+		return -1;
+	}
+
+	if (!(flags & ZSOCK_MSG_PEEK)) {
+		net_pkt_unref(pkt);
+	} else {
+		net_pkt_cursor_init(pkt);
+	}
+
+	return recv_len;
+}
+
+static int zcan_getsockopt_ctx(struct net_context *ctx, int level, int optname,
+			       void *optval, socklen_t *optlen)
+{
+	if (!optval || !optlen) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	return sock_fd_op_vtable.getsockopt(ctx, level, optname,
+					    optval, optlen);
+}
+
+static int zcan_setsockopt_ctx(struct net_context *ctx, int level, int optname,
+			       const void *optval, socklen_t optlen)
+{
+	return sock_fd_op_vtable.setsockopt(ctx, level, optname,
+					    optval, optlen);
+}
+
+static ssize_t can_sock_read_vmeth(void *obj, void *buffer, size_t count)
+{
+	return zcan_recvfrom_ctx(obj, buffer, count, 0, NULL, 0);
+}
+
+static ssize_t can_sock_write_vmeth(void *obj, const void *buffer,
+				    size_t count)
+{
+	return zcan_sendto_ctx(obj, buffer, count, 0, NULL, 0);
+}
+
+static int can_sock_ioctl_vmeth(void *obj, unsigned int request, va_list args)
+{
+	return sock_fd_op_vtable.fd_vtable.ioctl(obj, request, args);
+}
+
+/*
+ * TODO: A CAN socket can be bound to a network device using SO_BINDTODEVICE.
+ */
+static int can_sock_bind_vmeth(void *obj, const struct sockaddr *addr,
+			       socklen_t addrlen)
+{
+	return zcan_bind_ctx(obj, addr, addrlen);
+}
+
+/* The connect() function is no longer necessary. */
+static int can_sock_connect_vmeth(void *obj, const struct sockaddr *addr,
+				  socklen_t addrlen)
+{
+	return 0;
+}
+
+/*
+ * The listen() and accept() functions are without any functionality,
+ * since the client-Server-Semantic is no longer present.
+ * When we use RAW-sockets we are sending unconnected packets.
+ */
+static int can_sock_listen_vmeth(void *obj, int backlog)
+{
+	return 0;
```

Comment:
Same problem as with packet sockets - it's error to call these functions on non-SOCK_STREAM.